### PR TITLE
Fix issue #804: Logging to same file from multiple processes misses messages

### LINKED ISSUE #804: Logging to same file from multiple processes misses messages
When multiple processes log to the same file using concurrentWrites="true", not all of the messages are written to the file. The concurrent write retry is not working. 

This happens on Windows 8.1 Pro.

The test program in [this repository](https://github.com/bhaeussermann/NLogConcurrencyTest) demonstrates the problem. The program creates three processes to write 100 messages each, for a total of 300 lines. The resulting log file contains less than 300 lines.

Here is the config that is used:

```
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      autoReload="true"
      throwExceptions="false"
      internalLogLevel="Debug" internalLogFile="c:\temp\nlog-internal.log" >

	<targets>
		<target name="logfile" xsi:type="File" fileName="${basedir}\log.txt" concurrentWrites="true" />
	</targets>

	<rules>
		<logger name="*" minlevel="Info" writeTo="logfile" />
	</rules>
</nlog>
```

The internal log (Debug level):

```
2015-07-20 09:28:21.4103 Debug Setting 'UppercaseLayoutRendererWrapper.uppercase' to 'true'
2015-07-20 09:28:21.4183 Debug Setting 'UppercaseLayoutRendererWrapper.uppercase' to 'true'
2015-07-20 09:28:21.4193 Debug Setting 'FileTarget.concurrentWrites' to 'true'
2015-07-20 09:28:21.4193 Info Adding target File Target[logfile]
2015-07-20 09:28:21.4193 Debug Registering target logfile: NLog.Targets.FileTarget
2015-07-20 09:28:21.4263 Debug Setting 'FileTarget.concurrentWrites' to 'true'
2015-07-20 09:28:21.4263 Info Adding target File Target[logfile]
2015-07-20 09:28:21.4193 Debug Targets:
2015-07-20 09:28:21.4363 Info File Target[logfile]
2015-07-20 09:28:21.4363 Debug Rules:
2015-07-20 09:28:21.4383 Info logNamePattern: (:All) levels: [ ] appendTo: [ logfile ]
2015-07-20 09:28:21.4383 Debug --- End of NLog configuration dump ---
2015-07-20 09:28:21.4363 Info Watching path 'C:\Projects\AdHoc\NLogConcurrencyTest\NLogConcurrencyTest\bin\Debug' filter 'NLog.config' for changes.
2015-07-20 09:28:21.4383 Info Watching path 'C:\Projects\AdHoc\NLogConcurrencyTest\NLogConcurrencyTest\bin\Debug' filter 'NLog.config' for changes.
2015-07-20 09:28:21.4514 Info logNamePattern: (:All) levels: [ ] appendTo: [ logfile ]
2015-07-20 09:28:21.4514 Debug --- End of NLog configuration dump ---
2015-07-20 09:28:21.4564 Debug Error => logfile
2015-07-20 09:28:21.4564 Debug Fatal => logfile
2015-07-20 09:28:21.4514 Info Found 14 configuration items
2015-07-20 09:28:21.4744 Debug Targets for LogRunner.Program by level:
2015-07-20 09:28:21.4744 Debug Trace =>
2015-07-20 09:28:21.4744 Debug Debug =>
2015-07-20 09:28:21.4744 Debug Info => logfile
2015-07-20 09:28:21.4834 Debug Warn => logfile
2015-07-20 09:28:21.4834 Debug Error => logfile
2015-07-20 09:28:21.4834 Debug Fatal => logfile
2015-07-20 09:28:34.9355 Info Shutting down logging...
2015-07-20 09:28:34.9355 Info Stopping file watching for path 'C:\Projects\AdHoc\NLogConcurrencyTest\NLogConcurrencyTest\bin\Debug' filter 'NLog.config'
2015-07-20 09:28:34.9355 Info Closing old configuration.
2015-07-20 09:28:34.9725 Debug Closing logging configuration...
2015-07-20 09:28:34.9755 Info Stopping file watching for path 'C:\Projects\AdHoc\NLogConcurrencyTest\NLogConcurrencyTest\bin\Debug' filter 'NLog.config'
2015-07-20 09:28:34.9845 Info Closing old configuration.
2015-07-20 09:28:35.0085 Debug Closing logging configuration...
2015-07-20 09:28:35.0135 Debug Finished closing logging configuration.
2015-07-20 09:28:35.0135 Info Logger has been shut down.
2015-07-20 09:28:35.0696 Info Shutting down logging...
2015-07-20 09:28:35.0696 Info Stopping file watching for path 'C:\Projects\AdHoc\NLogConcurrencyTest\NLogConcurrencyTest\bin\Debug' filter 'NLog.config'
2015-07-20 09:28:35.0777 Info Closing old configuration.
2015-07-20 09:28:35.0927 Debug Closing logging configuration...
2015-07-20 09:28:35.0927 Debug Finished closing logging configuration.
2015-07-20 09:28:35.0927 Info Logger has been shut down.
```

### PR DESCRIPTION
**Pull request by bhaeussermann**
Fixes #804 

The concurrent write retry was not working for the RetryingMultiProcessFileAppender.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks! 

Is there a reason to rename the parameter? I think atomic should be the result of "allowConcurrentWrite" and so the name "allowConcurrentWrite" should be OK.

File: src/NLog/Internal/FileAppenders/BaseFileAppender.cs [hunk 1]
```diff
@@ -177,12 +177,12 @@ protected FileStream CreateFileStream(bool allowConcurrentWrite)
                         }
 
                         Directory.CreateDirectory(Path.GetDirectoryName(this.FileName));
-                        return this.TryCreateFileStream(allowConcurrentWrite);
+                        return this.TryCreateFileStream(allowAtomicConcurrentWrite);
                     }
                 }
                 catch (IOException)
                 {
-                    if (!this.CreateFileParameters.ConcurrentWrites || !allowConcurrentWrite || i + 1 == this.CreateFileParameters.ConcurrentWriteAttempts)
```
Comment (304NotModified) [inline src/NLog/Internal/FileAppenders/BaseFileAppender.cs]: Please add some comments why we should rethrow the exception when `allowConcurrentWrite=false`

Comment (bhaeussermann) [inline src/NLog/Internal/FileAppenders/BaseFileAppender.cs]: The `allowConcurrentWrite` parameter must apply only to how the stream is created when `TryCreateFileStream()` is called (that is, if true, then multiple processes may have an open stream for the file at the same time). 

The parameter must not affect the concurrent write retry. Write retry must still work when `allowConcurrentWrite = false`. Hence, the exception must not be thrown in this case.

Comment (304NotModified) [inline src/NLog/Internal/FileAppenders/BaseFileAppender.cs]: Thanks, can you add this as code comments? :)

**Comment by bhaeussermann** [comment]
The reason I've renamed the parameter was to make a clear distinction between the `allowConcurrentWrite` parameter and `CreateFileParameters.ConcurrentWrites`. The former value controls how the stream is created in `TryCreateFileStream()` (can multiple processes concurrently have an open stream to the file?), and the latter controls if write retries should be performed.

The two values have very similar names, yet they control two very different functions. In fact, this is probably the reason the bug manifested in the first place. 

The name `allowAtomicConcurrentWrite` was supposed to indicate that the value operates on a much lower level. Perhaps there's a better name. I'm currently thinking of `allowConcurrentStreams`.

What are your thoughts?

**Comment by 304NotModified** [comment]
Naming can be tricky. I see that the "allowAtomicConcurrentWrite" narrows down to:

- `FileShare.ReadWrite`
- `Win32FileNativeMethods.FILE_SHARE_WRITE`

Maybe something like "allowFileSharedWriting" is better? If we rename it, we should also rename it in the methods:

- `TryCreateFileStream`
- `WindowsCreateFile`

**Comment by 304NotModified** [comment]
PS the missing docs of `ICreateFileParameters` weren't helping. I added the xmldocs in https://github.com/NLog/NLog/pull/809

**Comment by 304NotModified** [comment]
Thanks!

**Pull request merged**